packages/block-library/src/query-loop/index.php

```diff
@@ -0,0 +1,58 @@
+<?php
+/**
+ * Server-side rendering of the `core/query-loop` block.
+ *
+ * @package WordPress
+ */
+
+/**
+ * Renders the `core/query-loop` block on the server.
+ *
+ * @param array    $attributes Block attributes.
+ * @param string   $content    Block default content.
+ * @param WP_Block $block      Block instance.
+ *
+ * @return string Returns the output of the query, structured using the layout defined by the block's inner blocks.
+ */
+function render_block_core_query_loop( $attributes, $content, $block ) {
+	$page_key = 'query-' . $block->context['id'] . '-page';
+	$page     = empty( $_GET[ $page_key ] ) ? 1 : $_GET[ $page_key ];
+	$posts    = get_posts(
+		array(
+			'post_type'      => 'post',
+			'posts_per_page' => $block->context['query']['per_page'],
+			'offset'         => $block->context['query']['per_page'] * ( $page - 1 ) + $block->context['query']['offset'],
+		)
+	);
+
+	$content = '';
+	foreach ( $posts as $post ) {
+		$content .= (
+			new WP_Block(
+				$block->parsed_block,
+				array_merge(
+					$block->available_context ? $block->available_context : array(),
```

Comment:
Why can't `available_context` be public?

The `context` registration doesn't become meaningless; they are just used for different things.

>Avoiding to be recreating the rendering, somehow allowing for a continuation of the normal render flow of inner blocks, simulated as if (or in reality of) these being inner blocks of the block

I don't see how we would clone the blocks with different contexts for every post.

>Some expression of a block like query loop to be able to receive all context (e.g. "context": "*", "context": true)

That would make the block rerender whenever unrelated context changes just to hack around making `available_context` public.

>Not passing along the ancestry context

We need it.

Reply:
> The `context` registration doesn't become meaningless; they are just used for different things.

Why should we expect a developer to implement `context`, if they could save themselves the effort by opting not to, and still have access to the same values?

Reply:
Because `context` subscribes the block to rerender when one of those values change.

In the future, it can also be used to scope blocks so that they can only be inserted inside blocks where the required context is present.

Reply:
> Because `context` subscribes the block to rerender when one of those values change.

But that's only relevant for the client-side, and still doesn't give us any guarantees that `context` will be provided as an accurate set of properties, especially if they're not using the context values in the editor and have no need for it. It also introduces an inconsistency where a server-side block instance has an additional property (`available_context`) which has no equivalent in the browser.

Reply:
Maybe there's something to explore here as an equivalent to React's [Portal](https://reactjs.org/docs/portals.html), in how it allows for the creation of a new tree of elements while retaining all of the existing context.

Reply:
It's not about retaining the context. It's about programmatically changing it.

I see the inconsistency there, but what problems do you think it could cause?

Reply:
> Because `context` subscribes the block to rerender when one of those values change.

That's an implementation detail about how we interoperate with React, not a reason why we really care for `context` to be a thing at a framework level. My prior understanding of why we wanted for `context` to be known is so that we can statically analyze the context needs of a block, in case we want to make further enhancements around block context needs, etc. We can't do this if we have no guarantees that context is provided.

Reply:
>In the future, it can also be used to scope blocks so that they can only be inserted inside blocks where the required context is present.

Yeah, static analysis is useful for things like that. I don't see how exposing `available_context` for advanced use cases detracts from that.

Reply:
> > In the future, it can also be used to scope blocks so that they can only be inserted inside blocks where the required context is present.
> 
> Yeah, static analysis is useful for things like that. I don't see how exposing `available_context` for advanced use cases detracts from that.

Because we have no guarantees they won't just use `available_context` for all context values. It's one less step they can save. There's an incentive to avoid effort. 

https://github.com/WordPress/gutenberg/pull/22199#discussion_r424068045

Reply:
That would be going out of their way to do things the wrong and undocumented way, and in that case, they would lose the benefits of static analysis. I don't think it's a problem.

Reply:
It's not going out of their way if it's the default state of inaction. I think it would be quite probable that one could stumble their way to `$block->available_context`, see some values they want to use, and never even realize that a `"context"` block type property even exists. Especially for things like `postId` and `postType`.

>they would lose the benefits of static analysis

What are these benefits?

Related to this, part of the question to me as well is: _Should_ we really care about `context`?

I'm not comfortable to just push forward with something like #22334, since I think it defers answering fundamental questions about what how we want context to work, in a way which doesn't oblige us to ever revisit it.

Reply:
Interesting challenge here. It's also related somehow to the fact that we store the template as "inner blocks" and not an attribute which for me is not ideal conceptually since these are not real inner blocks. It looks like we're doing a lot of magic here where IMO the solution should be something like:

```
do_blocks( template, context )
```

Where we set the post context as aa second argument. I'd even prefer if we avoid passing the parent context entirely, it's not clear at this point in which use-cases it's needed but if we ever have to support it `context = *` seems like a good approach to me.

Reply:
>Interesting challenge here. It's also related somehow to the fact that we store the template as "inner blocks" and not an attribute which for me is not ideal conceptually since these are not real inner blocks. It looks like we're doing a lot of magic here where IMO the solution should be something like:

That's not related to `available_context`. I tried storing the template in an attribute, and it was a lot more complicated and less performant.

>Where we set the post context as aa second argument. I'd even prefer if we avoid passing the parent context entirely, it's not clear at this point in which use-cases it's needed but if we ever have to support it context = * seems like a good approach to me.

You're right in that we might not even need to pass down `available_context` context. @aduth I'll update it so that we don't use `available_context` at all for now.